core/fetch/fetch.go

```diff
@@ -19,24 +19,37 @@ import (
 
 const heightPollingPeriod = 3 * time.Second
 
-var (
-	generatorHeight          uint64
-	generatorHeightFetchedAt time.Time
-	generatorLock            sync.Mutex
-)
+// New initializes a new Replicator to replicate blocks from the
+// Chain Core specified by peer. It immediately begins polling for
+// the peer's blockchain height, and will stop only when ctx is
+// cancelled. To begin replicating blocks, the caller must call
+// Fetch.
+func New(ctx context.Context, peer *rpc.Client) *Replicator {
+	rep := &Replicator{peer: peer}
 
-func GeneratorHeight() (uint64, time.Time) {
-	generatorLock.Lock()
-	h := generatorHeight
-	t := generatorHeightFetchedAt
-	generatorLock.Unlock()
-	return h, t
+	// Fetch the generator height periodically.
+	go rep.pollGeneratorHeight(ctx)
```

Comment:
Yeah, that's totally reasonable. I was also considering passing the ctx into all RunOptions?